src/Init/Init_FFTW.cpp

```diff
@@ -39,49 +83,95 @@ void Init_FFTW()
 
    if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ... ", __FUNCTION__ );
 
-// create plans for calculating the power spectrum
-#  ifdef SERIAL
-   FFTW_Plan_PS = rfftw3d_create_plan( NX0_TOT[2], NX0_TOT[1], NX0_TOT[0], FFTW_REAL_TO_COMPLEX,
-                                       FFTW_ESTIMATE | FFTW_IN_PLACE );
-#  else
-   FFTW_Plan_PS = rfftw3d_mpi_create_plan( MPI_COMM_WORLD, NX0_TOT[2], NX0_TOT[1], NX0_TOT[0],
-                                           FFTW_REAL_TO_COMPLEX, FFTW_ESTIMATE );
-#  endif
+#  ifdef SUPPORT_FFTW3
+   FFTW3_Double_OMP_Enabled  = false;
+   FFTW3_Single_OMP_Enabled = false;
 
 
-#  ifdef GRAVITY
+#  ifdef OPENMP
+
+#  ifndef SERIAL
+// check the level of MPI thread support
+   int MPI_Thread_Status;
+   MPI_Query_thread( &MPI_Thread_Status );
+
+// enable multithreading if possible
+   FFTW3_Double_OMP_Enabled = MPI_Thread_Status >= MPI_THREAD_FUNNELED;
+   FFTW3_Single_OMP_Enabled = FFTW3_Double_OMP_Enabled;
+#  else // # ifndef SERIAL
+
+// always enable multithreading in serial mode with openmp
+   FFTW3_Double_OMP_Enabled  = true;
+   FFTW3_Single_OMP_Enabled = true;
+#  endif // # ifndef SERIAL ... # else
+
+// initialise fftw multithreading
+   if (FFTW3_Double_OMP_Enabled) {
+      FFTW3_Double_OMP_Enabled = fftw_init_threads();
+      FFTW3_Single_OMP_Enabled = fftwf_init_threads();
+   }
+#  endif // # ifdef OPENMP
+
+// initialise fftw mpi support
+#  ifndef SERIAL
+   fftw_mpi_init();
+   fftwf_mpi_init();
+#  endif // # ifndef SERIAL
+
+// tell all subsequent fftw3 planners to use OMP_NTHREAD threads
+#  ifdef OPENMP
+   if (FFTW3_Double_OMP_Enabled) fftw_plan_with_nthreads (OMP_NTHREAD);
+   if (FFTW3_Single_OMP_Enabled) fftwf_plan_with_nthreads(OMP_NTHREAD);
+#  endif // # ifdef OPENMP
+#  endif // # ifdef SUPPORT_FFTW3
+
+
+
+// determine the FFT size for the power spectrum
+   int PS_FFT_Size[3]      = { NX0_TOT[0], NX0_TOT[1], NX0_TOT[2] };
+
 // determine the FFT size for the self-gravity solver
-   int FFT_Size[3] = { NX0_TOT[0], NX0_TOT[1], NX0_TOT[2] };
+#  ifdef GRAVITY
+   int Gravity_FFT_Size[3] = { NX0_TOT[0], NX0_TOT[1], NX0_TOT[2] };
 
 // the zero-padding method is adopted for the isolated BC.
    if ( OPT__BC_POT == BC_POT_ISOLATED )
-      for (int d=0; d<3; d++)    FFT_Size[d] *= 2;
+      for (int d=0; d<3; d++)    Gravity_FFT_Size[d] *= 2;
 
 // check
    if ( MPI_Rank == 0 )
    for (int d=0; d<3; d++)
    {
-      if ( FFT_Size[d] <= 0 )    Aux_Error( ERROR_INFO, "FFT_Size[%d] = %d < 0 !!\n", d, FFT_Size[d] );
+      if ( Gravity_FFT_Size[d] <= 0 )    Aux_Error( ERROR_INFO, "Gravity_FFT_Size[%d] = %d < 0 !!\n", d, Gravity_FFT_Size[d] );
    }
+#  endif // #  ifdef GRAVITY
 
+   real* PS   = NULL;
+   real* RhoK = NULL;
```

Comment:
Not unless we want to have two different calls to the `create_fftw_3d_xxx_plan` functions because their FFTW3 implementations require a pointer to the block of memory for the arrays. We could have something like 
`// create plans for power spectrum and the self-gravity solver
# if (SUPPORT_FFTW == FFTW3 ) 
   FFTW_Plan_PS      = create_fftw_3d_r2c_plan(PS_FFT_Size, PS);
#  ifdef GRAVITY
   FFTW_Plan_Poi     = create_fftw_3d_r2c_plan(Gravity_FFT_Size, RhoK);
   FFTW_Plan_Poi_Inv = create_fftw_3d_c2r_plan(Gravity_FFT_Size, RhoK);
#  endif // # ifdef GRAVITY`
# else // # if (SUPPORT_FFTW == FFTW3 ) 
   FFTW_Plan_PS      = create_fftw_3d_r2c_plan(PS_FFT_Size, NULL);
#  ifdef GRAVITY
   FFTW_Plan_Poi     = create_fftw_3d_r2c_plan(Gravity_FFT_Size, NULL);
   FFTW_Plan_Poi_Inv = create_fftw_3d_c2r_plan(Gravity_FFT_Size, NULL);
#  endif // # ifdef GRAVITY`
# endif // # if (SUPPORT_FFTW == FFTW3 ) `
But I think this code is harder to read.

Reply:
I misread the code. You are right.